spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/config/DataFlowServerConfigurationTests.java

```diff
@@ -51,118 +47,102 @@
-import org.springframework.core.env.StandardEnvironment;
+import org.springframework.core.NestedExceptionUtils;
 import org.springframework.hateoas.config.EnableHypermediaSupport;
 import org.springframework.security.authentication.AuthenticationManager;
 
-import static org.junit.Assert.assertFalse;
-import static org.junit.Assert.assertNotNull;
-import static org.junit.Assert.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertInstanceOf;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 import static org.mockito.Mockito.mock;
 
 /**
  * @author Glenn Renfro
  * @author Ilayaperumal Gopinathan
  * @author Gunnar Hillert
+ * @author Michael Wirth
  */
 public class DataFlowServerConfigurationTests {
 
-	private AnnotationConfigApplicationContext context;
+	private ApplicationContextRunner contextRunner;
 
-	private ConfigurableEnvironment environment;
-
-	private MutablePropertySources propertySources;
-
-	@Before
+	@BeforeEach
 	public void setup() {
-		context = new AnnotationConfigApplicationContext();
-		context.setId("testDataFlowConfig");
-		context.register(DataFlowServerConfigurationTests.TestConfiguration.class,
-				SecurityAutoConfiguration.class, DataFlowServerAutoConfiguration.class,
-				DataFlowControllerAutoConfiguration.class, DataSourceAutoConfiguration.class,
-				DataFlowServerConfiguration.class, PropertyPlaceholderAutoConfiguration.class,
-				RestTemplateAutoConfiguration.class, HibernateJpaAutoConfiguration.class, WebConfiguration.class,
-				SchedulerConfiguration.class, JacksonAutoConfiguration.class, SimpleTaskAutoConfiguration.class,
-				ResourceLoadingAutoConfiguration.class, ComposedTaskRunnerConfigurationProperties.class);
-		environment = new StandardEnvironment();
-		propertySources = environment.getPropertySources();
-	}
-
-	@After
-	public void teardown() {
-		if (context != null) {
-			context.close();
-		}
+		contextRunner = new ApplicationContextRunner()
+				.withAllowBeanDefinitionOverriding(true)
+				.withUserConfiguration(DataFlowServerConfigurationTests.TestConfiguration.class,
+						SecurityAutoConfiguration.class, DataFlowServerAutoConfiguration.class,
+						DataFlowControllerAutoConfiguration.class, DataSourceAutoConfiguration.class,
+						DataFlowServerConfiguration.class, PropertyPlaceholderAutoConfiguration.class,
+						RestTemplateAutoConfiguration.class, HibernateJpaAutoConfiguration.class,
+						SchedulerConfiguration.class, JacksonAutoConfiguration.class, SimpleTaskAutoConfiguration.class,
+						ResourceLoadingAutoConfiguration.class, ComposedTaskRunnerConfigurationProperties.class);
 	}
 
 	/**
 	 * Verify that embedded server starts if h2 url is specified with default properties.
 	 */
 	@Test
-	@Ignore
 	public void testStartEmbeddedH2Server() {
```

Comment:
I am going to share w/ you some of the beauty of the ApplicationContextRunner. One thing that it lets us do is really break apart the rest of the system and focus solely on the tiny piece of autoconfig we are testing. You will notice this current test has to bring in the kitchen sink to startup as it tests the H2Server startup + Dataflow server. 

Take a look at this instead: 
```java
/**
 * Tests for {@link H2ServerConfiguration}.
 *
 * @author Michael Wirth
 */
public class H2ServerConfigurationTests {

	private ApplicationContextRunner runner = new ApplicationContextRunner()
			.withConfiguration(AutoConfigurations.of(PropertyPlaceholderAutoConfiguration.class, H2ServerConfiguration.class));

	@Test
	void serverStartsWhenUrlIsH2AndEmbeddedPropertyTrue() {
		runner.withPropertyValues(
						"spring.datasource.url=jdbc:h2:tcp://localhost:19092/mem:dataflow",
						"spring.dataflow.embedded.database.enabled=true")
				.run(context -> assertThat(context)
						.hasNotFailed()
						.hasBean("h2TcpServer")
						.getBean("h2TcpServer", Server.class)
						.satisfies((server) -> {
							assertThat(server.isRunning(false)).isTrue();
							assertThat(server.getPort()).isEqualTo(19092);
						}));
	}

	@Test
	void serverStartsWhenUrlIsH2AndEmbeddedPropertyNotSet() {
		runner.withPropertyValues(
						"spring.datasource.url=jdbc:h2:tcp://localhost:19092/mem:dataflow")
				.run(context -> assertThat(context)
						.hasNotFailed()
						.hasBean("h2TcpServer")
						.getBean("h2TcpServer", Server.class)
						.satisfies((server) -> {
							assertThat(server.isRunning(false)).isTrue();
							assertThat(server.getPort()).isEqualTo(19092);
						}));
	}

	@Test
	void serverStopsWhenContextClosed() {
		runner.withPropertyValues(
						"spring.datasource.url=jdbc:h2:tcp://localhost:19092/mem:dataflow")
				.run(context -> {
					assertThat(context).hasNotFailed().hasBean("h2TcpServer");
					Server server = context.getBean("h2TcpServer", Server.class);
					assertThat(server.isRunning(false)).isTrue();
					context.close();
					assertThat(server.isRunning(false)).isFalse();
				});
	}

	@Test
	void serverDoesNotStartByDefault() {
		runner.run(context -> assertThat(context).hasNotFailed().doesNotHaveBean("h2TcpServer"));
	}

	@Test
	void serverDoesNotStartWhenUrlIsNotH2() {
		// TODO
	}

	@Test
	void serverDoesNotStartWhenEmbeddedPropertyFalse() {
		// TODO
	}

	@Test
	void serverDoesNotStartWhenUrlIsH2ButInvalidForm() {
		// TODO (may have more than one of these - 1 for port bad, 1 for wrong parts in url,, etc..)
	}

	@Test
	void serverDoesNotStartWhenH2NotOnClasspath() {
		// look at withFilteredClassLoader(new FilteredClassloader(Server.class))
	}
}
```

Of course testing H2 + the full app is needed but we can have a single (+) and (-) h2 case in the DF server tests.

Reply:
If later you find yourself wanting to make the above assertions more DRY, you can do things like this: 

```java
	@Test
	void serverStartsWhenUrlIsH2AndEmbeddedPropertyTrue() {
		runner.withPropertyValues(
						"spring.datasource.url=jdbc:h2:tcp://localhost:19092/mem:dataflow")
				.run(assertServerStarted(19092, null));
	}

	@Test
	void serverStartsPlusCheckiingOtherFancyThings{
		runner.withPropertyValues(
                                           "spring.datasource.url=jdbc:h2:tcp://localhost:19092/mem:dataflow")
				.run(assertServerStarted(19092, (server) -> {
                                                assertThat(server).blahBlahBlah().youGetTheIdea();
                                           ));
	}

	private ContextConsumer<AssertableApplicationContext> assertServerStarted(int port, Consumer<Server> serverConsumer) {
		return (context) -> {
			assertThat(context).hasNotFailed().hasBean("h2TcpServer");
			Server server = context.getBean("h2TcpServer", Server.class);
			assertThat(server.isRunning(false)).isTrue();
			assertThat(server.getPort()).isEqualTo(port);
			if (serverConsumer != null) {
				serverConsumer.accept(server);
			}
		};
	}
```

Reply:
Cools testing feature - thanks for the advise

Reply:
👍🏻 you are more than welcome. It is definitely one of my favorites.